File: minesweeper/minesweeper.py
import random

def get_number(a, b):
    while True:
        try:
            val = int(input(f"podaj liczbe z zakresu {a} - {b}: "))

        except ValueError:
            print("to nie jest liczba ")

        else:
            if val in range(a,b+1):
                return val

def lay_mines(noOfMines, x, y):
    arr = set()
    while len(arr) < noOfMines:
        i = random.randrange(x)
        j = random.randrange(y)
        arr.add((i,j))
    
    return arr

def number_of_neighboring_mines(coord, coordsList):
    noOfMines = 0
    for i in coordsList:
        for a in range(-1,2):
            for b in range(-1,2):
                if (coord[0] - a, coord[1] - b) == i:
                    noOfMines += 1
    
    return noOfMines

# def number_of_neighboring_mines2(coord, minesList):
#     noOfMines = 0
#     j = coord[0]
#     i = coord[1]
#     for x in [(i-1, j-1), (i-1, j), (i-1, j+1),
#                 (i,j-1), (i, j+1),
#                 (i+1, j-1), (i+1, j), (i+1, j-1)]:
#         if x in minesList:
#             noOfMines += 1;
    
#     return noOfMines;

def create_board(x, y, mines, mine="*"):
    borad = []
    mask = []
    for i in range(x):
        row = []
        oRow = []
        for j in range(y):
            oRow.append("_")
            if (i,j) in mines:
                row.append(mine)
            else:
                row.append(number_of_neighboring_mines((i,j), mines))
        borad.append(row)
        mask.append(oRow)

    return (borad, mask)

def reveal_fields(x, y, board, mask, isFirst = True):

    if board[x][y] == "*" and isFirst == True:
        mask[x][y] = "*"
        return board, mask

    if board[x][y] > 0:
        mask[x][y] = "N"
        return board, mask
    
    mask[x][y] = "E"
    for a in range(-1, 2):
        for b in range(-1, 2):
            new_x, new_y = x + a, y + b
            if 0 <= new_x < len(board) and 0 <= new_y < len(board[0]) and mask[new_x][new_y] == "_":
                board, mask = reveal_fields(new_x, new_y, board, mask, isFirst = False)

    return board, mask

def print_board(board, mask):
    
    for row in range(len(board)):
        for el in range(len(board[row])):
            if mask[row][el] == "N":
                print(f"{board[row][el]:^3}", end="")
            elif mask[row][el] == "E":
                print("   ", end="")
            else:
                print(f"{mask[row][el]:^3}", end="")
        print()

def contiueCheck(mask):
    for row in mask:
        if "*" in row:
            print("\ngame over")
            return False
        
    for row in mask:
        if '_' in row:
             return True
    
    print("\nwygranko")
    return False

def sapper():
    x = 10
    y = 10
    noOfMines = 30
    mines = lay_mines(noOfMines, x, y)
    board, mask = create_board(x, y, mines)
    
    print_board(board, mask)

    isGameOn = True
    while isGameOn:
        coordX = get_number(1,x) - 1
        coordY = get_number(1,y) - 1
        board, mask = reveal_fields(coordX, coordY, board, mask)
        print_board(board, mask)
        isGameOn = contiueCheck(mask)

sapper()

File: minesweeper/minesweeper.java

package minesweeper;

import java.util.Scanner;
import java.util.Arrays;
import java.util.Random;

public class minesweeper {

    static int getNumber( int a, int b){
        Scanner scanner = new Scanner(System.in);
        int num = 0;  
        try {
            System.out.printf("Enter number in range %d to %d:\t", a, b-1);
            num = scanner.nextInt();
            if (num >= a && num <= b)
            return num;
            else{
                System.out.println("wrong number");   
                return getNumber(a, b);
            }
        } 
        catch (NumberFormatException e) {
            scanner.nextLine(); 
            throw new NumberFormatException("Invalid input. Please enter a valid integer.");
        }
    }

    static int[][] layMines(int rows, int cols, int noOfMines){
        int[][] arr = new int[noOfMines][2];
        Random rand = new Random();
        
        for (int i = 0; i < noOfMines; i++){
            int val1,val2;  
            do {
                val1 = rand.nextInt(rows);
                val2 = rand.nextInt(cols);
            } while (!isUnique(arr, val1, val2));
            arr[i][0] = val1;
            arr[i][1] = val2;
        }
        return arr;
    }

    static boolean isUnique(int[][] unique, int val1, int val2) {
        for (int[] pair : unique) {
            if (pair[0] == val1 && pair[1] == val2) {
                return false;
            }
        }
        return true;
    }

    static int numberOfNeighboringMines(int x, int y, int[][] minesList){
        int count = 0;
        for (int i = -1; i < 2; i++){
            for (int j = -1; j < 2; j++){
                if (i == 0 && j == 0)
                    continue;
                int [] tempArr = {x + i, y + j};
                for(int[] row : minesList)
                    if (Arrays.equals(row, tempArr))
                        count ++;
            }
        }
        return count;
    }

    static String[][] createBoard(int rows, int cols, int[][] minesList){
        String[][] board = new String[rows][cols];
        for (int[] coord : minesList){
            board[coord[0]][coord[1]] = "*";
        }

        for (int i = 0; i < rows; i++){
            for (int j = 0; j < cols; j++){
                if (board[i][j] != "*"){
                    Integer y = numberOfNeighboringMines(i, j, minesList);
                    board[i][j] = y.toString();
                }
            }
        }
        return board;
    }

    static String[][] createMask(int rows, int cols){
        String[][] mask = new String[rows][cols];

        for (int i = 0; i < rows; i++){
            Arrays.fill(mask[i], "-");
        }
        return mask;

    }

    static String[][] revealFields(int x, int y, String[][] board, String[][] mask) {
        if (mask[x][y].equals("-")) { // Check if the field is unrevealed
            if (board[x][y].equals("*"))
                mask[x][y] = "M";
            else
                mask[x][y] = "U";
            
            if (board[x][y].equals("0")) { // If the field is equal to "0"
                for (int i = -1; i < 2; i++) {
                    for (int j = -1; j < 2; j++) {
                        if (i == 0 && j == 0)
                            continue;
                        
                        int newX = x + i;
                        int newY = y + j;
                        
                        if (newX >= 0 && newX < board.length && newY >= 0 && newY < board[0].length) {
                            mask = revealFields(newX, newY, board, mask);
                        }
                    }
                }
            }
        }
        
        return mask;
    }
    
    static void printMap(String[][] board, String[][] mask) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (mask[i][j] == "U" || mask[i][j] == "M")
                    System.out.print(board[i][j] + " ");
                else
                    System.out.print(mask[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println();
    }

    static boolean isGameOn(String[][] mask){
        for (int i = 0; i < mask.length; i++) {
            for (int j = 0; j < mask[i].length; j++) {
                if (mask[i][j].equals("M")){
                    System.out.println("GAME OVER");
                    return false;
                }
                if (mask[i][j].equals("-"))
                    return true;
            }
        }
        System.out.println("YOU WON");
        return false;
    }       

    static void game(int x, int y, int noOfMi){
        int [][] minArr = layMines(x, y, noOfMi);
        String[][] board = createBoard(x, y, minArr);
        String[][] mask = createMask(x, y);
        printMap(board, mask);
        
        while (isGameOn(mask)){
            int coordX = getNumber(0, x);
            int coordY = getNumber(0, y);
            revealFields(coordX, coordY, board, mask);
            printMap(board, mask);
        }
    }
    public static void main(String[] args){
        game(10, 10, 10);
    }   
     
}